src/Travers.php

```diff
@@ -65,6 +89,38 @@ public function find($key)
         return $this->parseParam($this->parseKeys($key), $this->tree);
     }
 
+    /**
+     * @param string $key
+     * @param array  $tree
+     *
+     * @return array|mixed|null
+     * @throws BranchNotFoundException
+     */
+    protected function removeParam($key, array $tree)
+    {
+        $tmp       = &$tree;
+        $keysArray = $this->parseKeys($key);
+
+        while (count($keysArray) > 1) {
+            $key = array_shift($keysArray);
+
+            if (!is_array($tmp) || !array_key_exists($key, $tmp)) {
+                return $this->shouldFail() ? $this->handleFailure($key) : $tmp;
+            }
+
+            $tmp = &$tmp[$key];
+        }
+
+        $key = array_shift($keysArray);
```

Comment:
This is basically the exact same code block as the one in the while loop. The while loop can be optimized.